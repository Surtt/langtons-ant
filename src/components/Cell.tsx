import React, { FC } from 'react';

const Cell: FC = () => {
  return (
    <td className="cell">
      <div className="box">
        <div className="content"></div>
      </div>
    </td>
  );
};

export default Cell;
