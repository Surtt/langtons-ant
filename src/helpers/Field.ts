type Cell = false | true;
type Line = Cell[];
export type Field = Line[];

export const generateField = (size: number, state: Cell): Field =>
  new Array<boolean[]>(size)
    .fill(new Array<boolean>(size))
    .map(() => new Array<boolean>(size).fill(state));

// export const toggleCell = (cell: Cell, coords: Coords): Cell => { };
