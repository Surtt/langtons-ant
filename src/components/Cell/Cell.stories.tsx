import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Cell, CellProps } from './Cell';
import { BoardTable } from '../Board/Board';

export default {
  title: 'Board/Cell',
  component: Cell,
  decorators: [
    (Story) => (
      <BoardTable style={{ width: '2vw', height: '2vw' }}>
        <tr>
          <Story />
        </tr>
      </BoardTable>
    ),
  ],
} as Meta;

const Template: Story<CellProps> = (args) => <Cell {...args} />;

export const CellExample = Template.bind({});
CellExample.args = {
  cell: false,
};
