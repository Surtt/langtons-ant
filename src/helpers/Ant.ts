import { Cell } from '@/helpers/Field';
export type Coords = [number, number];

type RandomGenerator<R = number> = (
  min: number,
  max: number,
  rotation?: number
) => R;

export type Ant = {
  xAnt: number;
  yAnt: number;
  rotation: number;
};

export const generateNumber: RandomGenerator = (min, max) => {
  return Math.floor(Math.random() * (max - min)) + min;
};

export const generateAntPosition: RandomGenerator<Ant> = (
  min: number,
  max: number,
  rotation = 0
) => {
  const xAnt = generateNumber(min, max);
  const yAnt = generateNumber(min, max);
  return {
    xAnt,
    yAnt,
    rotation,
  };
};

export const changeAntState = (
  rotation: number,
  cell: Cell
): {
  xAnt: number;
  yAnt: number;
} => {
  const value = { xAnt: 0, yAnt: 0 };
  switch (rotation) {
    case 90:
      value.xAnt++;
      break;
    case 180:
      value.yAnt--;
      break;
    case 270:
      value.xAnt--;
      break;
    default:
      value.yAnt++;
      break;
  }
  if (cell) {
    value.xAnt = -value.xAnt;
    value.yAnt = -value.yAnt;
  }
  return value;
};

export const newRotation = (rotation: number, cell: Cell): number => {
  let newRotation = rotation + 90;
  if (cell) {
    newRotation += 180;
  }
  if (newRotation >= 360) {
    newRotation -= 360;
  }
  return newRotation;
};
