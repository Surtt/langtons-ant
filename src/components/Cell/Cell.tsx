import React, { FC } from 'react';
import styled from '@emotion/styled';
import { css } from '@emotion/react';
import { Ant } from '@/helpers/Ant';
import img from './ant.png';

export interface CellProps {
  /**
   * Cell data
   */
  cell?: boolean;
  ant?: Ant;
}

export const Cell: FC<CellProps> = ({ ant, cell }) => {
  // console.log(cell);
  const isColored = cell ? 'orange' : '';
  // const hasAnt = ant ? (ant.rotation !== 0 ? `ant${ant.rotation}` : 'ant') : '';
  // const styleAntOrNot = `${hasAnt}${isColored}`;
  return (
    <CellTd role="cell">
      {/* <Box>
        {ant ? <AntCell /> : <EmptyCell />}
        {cell ? <FilledCell /> : <EmptyCell />}
      </Box> */}
      {/* <Box>{ant ? <AntCell /> : cell ? <FilledCell /> : <EmptyCell />}</Box> */}
      <Box>
        {ant ? (
          <AntCell rotate={ant.rotation} />
        ) : (
          <CellContent className={isColored} />
        )}
      </Box>
    </CellTd>
  );
};

const CellTd = styled.td`
  border: 1px solid #337ab7;
`;

const Box = styled.div`
  position: relative;
  width: 100%;
  &:before {
    content: '';
    display: block;
    padding-top: 100%;
  }
`;

const CellContent = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
`;

// const FilledCell = styled(EmptyCell)`
//   background-color: orange;
// `;

interface RotateProps {
  rotate: number;
}
const dynamicStyleAnt = ({ rotate }: RotateProps) =>
  css`
    background: url(${img}) no-repeat;
    background-size: contain;
    background-position: center;
    transform: rotate(${rotate}deg);
  `;

const AntCell = styled(CellContent)`
  ${dynamicStyleAnt};
`;
