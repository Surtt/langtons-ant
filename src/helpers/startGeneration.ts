import { changeAntState, newRotation, Ant } from './Ant';
import { Field, toggleCellOnTheField } from './Field';
export interface GeneratingValues {
  field: Field;
  ant: Ant;
  count: number;
}

const isFieldBorder = (field: Field, ant: Ant) => {
  const borderY = field.length - 1;
  const borderX = field[0].length - 1;
  return (
    ant.yAnt === 0 ||
    ant.yAnt === borderY ||
    ant.xAnt === 0 ||
    ant.xAnt === borderX
  );
};

export const startGeneration = (
  field: Field,
  ant: Ant,
  count: number
): GeneratingValues => {
  const cnt = count + 1;
  const { xAnt, yAnt, rotation } = ant;

  const movement = changeAntState(rotation, field[xAnt][yAnt]);
  const updateRotation = newRotation(rotation, field[xAnt][yAnt]);
  field = toggleCellOnTheField(field, [xAnt, yAnt]);
  movement.xAnt += xAnt;
  movement.yAnt += yAnt;
  const newAnt = {
    xAnt: movement.xAnt,
    yAnt: movement.yAnt,
    rotation: updateRotation,
  };

  if (isFieldBorder(field, ant)) {
    const length = field.length;
    field.unshift(new Array<boolean>(length).fill(false));
    field.push(new Array<boolean>(length).fill(false));
    field.forEach((item) => {
      item.unshift(false);
      item.push(false);
    });
    if (newAnt) {
      newAnt.xAnt++;
      newAnt.yAnt++;
    }
  }

  return {
    field: [...field],
    ant: newAnt,
    count: cnt,
  };
};
