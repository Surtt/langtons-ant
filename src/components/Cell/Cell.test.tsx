import React from 'react';
import { render, screen } from '@testing-library/react';
import { Cell } from './Cell';
import { Cell as CellType } from '@/helpers/Field';

const cell: CellType = false;
const ant = {
  xAnt: 0,
  yAnt: 0,
  rotation: 0,
};

it('Renders Cell', () => {
  const { asFragment } = render(<Cell ant={ant} cell={cell} />);
  expect(asFragment()).toMatchSnapshot();
});

it('renders Line role', () => {
  render(<Cell ant={ant} cell={cell} />);
  expect(screen.getByRole('cell')).toBeInTheDocument();
});
