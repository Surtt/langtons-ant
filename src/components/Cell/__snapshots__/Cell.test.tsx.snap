// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Cell 1`] = `
<DocumentFragment>
  <div
    class="css-1h0vbn9"
  >
    <div
      class="css-fzp5ig"
      rotate="0"
    />
  </div>
</DocumentFragment>
`;
